FLIGHT_DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,442 @@
+objects. There are two styles in use described in the next two sections.
+
+#### Stairway Application Context set to Spring Application Context
+
+Suppose you have an autowired class called` FooService`. How can you resolve FooService in
+your flight?
+
+You can set the Stairway application context to be the Spring Application Context like this:
+```java
+@Autowired
+ApplicationContext applicationContext; // Spring application context
+
+Stairway.newBuilder(
+  .applicationContext(applicationContext)
+  ...
+  ).build();
+```
+
+Then your flight constructors need code like this:
+```java
+public YourFlight(FlightMap inputParameters, Object applicationContext) {
+  super(inputParameters, applicationContext);
+
+  ApplicationContext springAppContext = (ApplicationContext)applicationContext;
+  FooService fooService = springAppContext.getBean(FooService.class);
+}
+```
+
+#### Resolve Beans in a "Bean Bag"
+
+An alternative to pulling beans from the Spring application context is to make a so-called
+"Bean Bag" class that resolves the beans and provides getters for use in flight
+constructors. The bean bag is used as the application context. It goes like this:
+
+Set up the bean bag class:
+
+```java
+@Component
+public class BeanBag {
+  private final FooService fooService;
+
+  @Autowired
+  public BeanBag(FooService fooService) {
+    this.fooService = fooService;
+  }
+
+  public FooService getFooService() {
+    return fooService;
+  }
+}
+```
+
+Use the bean bag class as the application context:
+
+```java
+@Autowired
+BeanBag beanBag;
+
+Stairway.newBuilder(
+  .applicationContext(beanBag)
+  ...
+  ).build();
+```
+
+Then your flight constructors look like this:
+
+```java
+  public YourFlight(FlightMap inputParameters, Object applicationContext) {
+    super(inputParameters, applicationContext);
+
+  BeanBag beanBag = (BeanBag)applicationContext;
+  FooService fooService = beanBag.getFooService();
+```
+
+## Dismal Failures <a name="dismalfailures"></a>
+
+Stairway uses the term _dismal failure_ to refer to a flight that has an error and then
+fails to successfully undo. It is the case where the saga transaction is not
+all-or-nothing. This is a serious issue, because the system is left in an unintended,
+potentially corrupt state. It probably requires human intervention to repair the problem.
+
+When this happens, Stairway writes a log message including the text
+`DISMAL FAILURE` that can be used for alerting.
+
+In our experience, there are two main causes of dismal failures:
+ 1. Expired retries
+ 1. Untested undo paths
+
+## Using FlightDebugInfo <a name="flightdebuginfo"></a>
+
+Stairway provides some help test flights. In a test environment, you can pass a
+`FlightDebugInfo` object in when you submit your flight. There are two types of behavior
+you can request.
+
+### Restart Every Step  <a name="restarteverystep"></a>
+
+You can request that Stairway restart your flight at each step boundary. That tests
```

Comment:
does it just restart once at each boundary?

Reply:
It restarts the flight every time it gets to the end of a step, going in either direction.